### PR TITLE
CP-14031: Update pool-join checks when management interface is on vlan.

### PR DESCRIPTION
**Pull request by sharady**

### CODE REVIEW

File: ocaml/xapi/xapi_pool.ml [hunk 1]
```diff
@@ -190,13 +190,29 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
...
 
-  let assert_only_physical_pifs () =
-    let non_physical_pifs = Db.PIF.get_refs_where ~__context ~expr:(
+  let allow_management_interface_on_vlan () =
+    (* Allow pool-join if joining host and the pool have management interface on same vlan *)
+    match Db.VLAN.get_all ~__context with
+    | [] -> false
+    | [vlan] ->
+      (try
+        if Db.PIF.get_VLAN ~__context ~self:(
```
Comment (lindig) [inline ocaml/xapi/xapi_pool.ml]: You can replace  `if <expr> then true else false` with `expr`. You might want to let-bind some values to structure the `expr`.

Comment (sharady) [reply ocaml/xapi/xapi_pool.ml]: Indeed we can remove that, Thanks.

File: ocaml/xapi/xapi_pool.ml [hunk 2]
```diff
@@ -190,13 +190,29 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
...
+    (* Allow pool-join if joining host and the pool have management interface on same vlan *)
+    match Db.VLAN.get_all ~__context with
+    | [] -> false
+    | [vlan] ->
+      (try
+        Db.PIF.get_VLAN ~__context ~self:(
+          Xapi_host.get_management_interface ~__context ~host:(Helpers.get_localhost ~__context)) =
+        Client.PIF.get_VLAN rpc session_id (
+          Client.Host.get_management_interface rpc session_id (get_master ~rpc ~session_id))
+       with e -> error "Caught %s: while checking joining host and the pool have management interface on same vlan" (ExnHelper.string_of_exn e);
```
Comment (lindig) [inline ocaml/xapi/xapi_pool.ml]: The error message is somewhat difficult to parse. How about: 

```
error "Caught '%s' while checking if joining host and the pool share a management interface on the same vlan"
```

or 
```
error "Caught '%s' while checking if joining host and the pool have a management interface on same vlan"
```

**Review by lindig — COMMENTED** [review]
This looks good. I suggest to update one error message because it is hard to parse and then the code can go in.

File: ocaml/xapi/xapi_pool.ml (hunk 2 above)
Comment (sharady) [reply ocaml/xapi/xapi_pool.ml]: Updated!!

File: ocaml/xapi/xapi_pool.ml [hunk 3]
```diff
@@ -190,13 +190,29 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
...
+    let non_physical_pifs_count = List.length (Db.PIF.get_refs_where ~__context ~expr:(
         Eq (Field "physical", Literal "false")
-      ) in
-    if non_physical_pifs <> [] then begin
-      error "The current host has network bonds, VLANs or tunnels: it cannot join a new pool";
-      raise (Api_errors.Server_error(Api_errors.pool_joining_host_must_only_have_physical_pifs, []))
+      )) in
+    (* Allow pool-join if host has only physical pifs or at most one vlan *)
+    if (non_physical_pifs_count > 1) || ((non_physical_pifs_count = 1) && not (allow_management_interface_on_vlan ())) then begin
+      error "The current host has network bonds, tunnels or multiple VLANs: it cannot join a new pool";
```
Comment (robhoes) [inline ocaml/xapi/xapi_pool.ml:214]: This error message, and the name+doc of the new API error, do not precisely describe the error condition. Even one VLAN is not allowed, if this VLAN is not used for management. A further constraint is that the management VLAN must match the one on the pool.

File: ocaml/xapi/xapi_pool.ml [hunk 4]
```diff
@@ -190,13 +190,29 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
       raise (Api_errors.Server_error(Api_errors.pool_joining_host_cannot_contain_shared_SRs, []))
     end in
 
-  let assert_only_physical_pifs () =
-    let non_physical_pifs = Db.PIF.get_refs_where ~__context ~expr:(
+  let allow_management_interface_on_vlan () =
+    (* Allow pool-join if joining host and the pool have management interface on same vlan *)
+    match Db.VLAN.get_all ~__context with
+    | [] -> false
+    | [vlan] ->
```
Comment (robhoes) [inline ocaml/xapi/xapi_pool.ml:197]: `vlan` us unused.

File: ocaml/xapi/xapi_pool.ml [hunk 5]
```diff
@@ -190,13 +190,29 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
...
+      (try
+        Db.PIF.get_VLAN ~__context ~self:(
```
Comment (robhoes) [inline ocaml/xapi/xapi_pool.ml:199]: I think that this allows the case where there is one VLAN, but the management interface is "physical", which we don't want to. In this case, the `PIF.VLAN` field is `-1`.

**Review by robhoes — COMMENTED** [review]
Sorry for the late review. I know this has already gone in, but I'd like my comments to be addressed in a further PR.

File: ocaml/xapi/xapi_pool.ml [hunk 6]
```diff
@@ -190,13 +190,29 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
...
         Eq (Field "physical", Literal "false")
-      ) in
-    if non_physical_pifs <> [] then begin
-      error "The current host has network bonds, VLANs or tunnels: it cannot join a new pool";
-      raise (Api_errors.Server_error(Api_errors.pool_joining_host_must_only_have_physical_pifs, []))
+      )) in
+    (* Allow pool-join if host has only physical pifs or at most one vlan *)
+    if (non_physical_pifs_count > 1) || ((non_physical_pifs_count = 1) && not (allow_management_interface_on_vlan ())) then begin
+      error "The current host has network bonds, tunnels or multiple VLANs: it cannot join a new pool";
+      raise (Api_errors.Server_error(Api_errors.pool_joining_host_must_only_have_physical_pifs_or_vlan_pif, []))
```
Comment (robhoes) [inline ocaml/xapi/xapi_pool.ml:215]: Let's split this out into a couple of more specific error codes for the following cases, so that it's clearer for clients the understand what is wrong:

- Management VLANs don't match.
- Joining host has bonds.
- Joining host has non-management VLANs.
- Joining host has tunnels.
